sklearn/calibration.py

```diff
@@ -887,6 +980,138 @@ def loss_grad(AB):
+
+    return logits
+
+
+def _temperature_scaling(predictions, labels, sample_weight=None, beta_0=1.0):
+    """Probability Calibration with temperature scaling
+
+    Parameters
+    ----------
+    predictions : ndarray of shape (n_samples,) or (n_samples, n_classes)
+        The output of `decision_function` or `predict_proba`. If the input
+        appears to be probabilities (i.e., values between 0 and 1 that sum to 1
+        across classes), it will be converted to logits using `np.log(p + eps)`.
+
+        Binary decision function outputs (1D) will be converted to two-class
+        logits of the form (-x, x). For shape (n_samples, 1), the same applies.
+
+    labels : ndarray of shape (n_samples,)
+        True labels for the samples.
+
+    sample_weight : array-like of shape (n_samples,), default=None
+        Sample weights. If None, then samples are equally weighted.
+
+    beta_0 : float, default=1.0
+       Initial inverse temperature to start the optimisation.
+
+    Returns
+    -------
+    beta : float
+        The optimised inverse temperature parameter for probability calibration,
+        with a value in the range (0, infinity).
+
+    References
+    ----------
+    On Calibration of Modern Neural Networks,
+    C. Guo, G. Pleiss, Y. Sun, & K. Q. Weinberger, ICML 2017.
+    """
+    check_consistent_length(predictions, labels)
+    logits = _standardize_decision_values(predictions)
+
+    xp, _ = get_namespace(logits, labels, sample_weight)
+    dtype_ = _find_matching_floating_dtype(logits, labels, sample_weight, xp=xp)
+    labels = column_or_1d(labels, dtype=dtype_)
+
+    if sample_weight is not None:
+        sample_weight = _check_sample_weight(sample_weight, labels, dtype=dtype_)
+        sample_weight = xp.asarray(sample_weight)
+
+    halfmulti_loss = HalfMultinomialLoss(
+        sample_weight=sample_weight, n_classes=logits.shape[1]
+    )
+
+    def beta_loss(beta=beta_0):
+        """Compute the negative log likelihood loss and its derivative
+            with respect  to temperature.
+
+        Parameters
+        ----------
+        beta : float
+            The current inverse temperature value during optimisation.
+
+        Returns
+        -------
+        negative_log_likelihood_loss : float
+            The negative log likelihood loss.
+
+        """
+        # Ensure raw_prediction has the same dtype as labels using .astype().
+        # Without this, dtype promotion rules differ across NumPy versions:
+        #
+        #   beta = np.float64(0)
+        #   logits = np.array([1, 2], dtype=np.float32)
+        #
+        #   result = beta * logits
+        #   - NumPy < 2: result.dtype is float32
+        #   - NumPy 2+:  result.dtype is float64
+        #
+        #  This can cause dtype mismatch errors downstream (e.g., buffer dtype).
+        raw_prediction = xp.astype(beta * logits, dtype_)
+
+        l = halfmulti_loss.loss(y_true=labels, raw_prediction=raw_prediction)
```

Comment:
Since the loss module is implemented in Cython will using the array api work here for example with PyTorch?

Reply:
No, it won't.

Reply:
Yes that makes sense. I think then in the initial inclusion of this feature we should avoid introducing the array api here.